lib/Service/ScrobblerService.php

```diff
@@ -0,0 +1,234 @@
+<?php declare(strict_types=1);
+
+/**
+ * ownCloud - Music app
+ *
+ * This file is licensed under the Affero General Public License version 3 or
+ * later. See the COPYING file.
+ *
+ * @author Matthew Wells
+ * @copyright Matthew Wells 2025
+ */
+
+namespace OCA\Music\Service;
+
+use DateTime;
+use OCA\Music\AppFramework\Core\Logger;
+use OCA\Music\BusinessLayer\TrackBusinessLayer;
+use OCA\Music\Db\Track;
+use OCP\IConfig;
+use OCP\IURLGenerator;
+use OCP\Security\ICrypto;
+
+class ScrobblerService
+{
+	private IConfig $config;
+
+	private Logger $logger;
+
+	private IURLGenerator $urlGenerator;
+
+	private TrackBusinessLayer $trackBusinessLayer;
+
+	private ?string $appName;
+
+	private ICrypto $crypto;
+
+	public const SCROBBLE_SERVICES = [
+		'lastfm' => [
+			'endpoint' => 'http://ws.audioscrobbler.com/2.0/',
+			'name' => 'Last.fm'
+		]
+	];
+
+	public function __construct(
+		IConfig $config,
+		Logger $logger,
+		IURLGenerator $urlGenerator,
+		TrackBusinessLayer $trackBusinessLayer,
+		ICrypto $crypto,
+		?string $appName = 'music'
+	) {
+		$this->config = $config;
+		$this->logger = $logger;
+		$this->urlGenerator = $urlGenerator;
+		$this->trackBusinessLayer = $trackBusinessLayer;
+		$this->crypto = $crypto;
+		$this->appName = $appName;
+	}
+
+	/**
+	 * @throws \Throwable when unable to generate or save a session
+	 */
+	public function generateSession(string $token, string $userId) : void {
+		$scrobbleService = $this->getApiService();
+		$ch = $this->makeCurlHandle($scrobbleService);
+		$params = $this->generateBaseMethodParams('auth.getSession');
+		$params['token'] = $token;
+		$params['api_sig'] = $this->generateSignature($params);
+		\curl_setopt($ch, \CURLOPT_POSTFIELDS, \http_build_query($params));
+		$sessionText = \curl_exec($ch);
+		$xml = \simplexml_load_string($sessionText);
+
+		$status = (string)$xml['status'];
+		if ($status !== 'ok') {
+			throw new \Exception((string)$xml->error, (int)$xml->error['code']);
+		}
+
+		try {
+			$encryptedKey = $this->crypto->encrypt(
+				(string)$xml->session->key,
+				$userId . $this->config->getSystemValue('secret')
+			);
+			$this->config->setUserValue($userId, $this->appName, 'scrobbleSessionKey', $encryptedKey);
+		} catch (\Throwable $e) {
+			$this->logger->error('Unable to save session key ' . $e->getMessage());
+			throw $e;
+		}
+	}
+
+	public function getApiKey() : ?string {
+		return $this->config->getSystemValue('music.scrobble_api_key', null);
```

Comment:
We already have the settings key `music.lastfm_api_key` so having another for the same purpose doesn't seem necessary. And if we later add support for Libre.fm, then we could add another key `music.librefm_api_key`. Then, ideally, the admin could configure even both keys at the same time, and each user would be allowed to select the scrobbling service they prefer.

To lay ground work for supporting multiple services like described above, also `music.scrobble_api_secret` should be `music.lastfm_api_secret`. Also `music.scrobble_api_service` is unnecessary since the availability of a scobbling service can be deduced from the availability of the API key *and* API secret.